python/cuml/feature_extraction/text.py

```diff
@@ -0,0 +1,603 @@
+                                 delimiter=self.delimiter)
+        return lambda doc: self._remove_stop_words(preprocess(doc))
+
+    def _get_stop_words(self):
+        """Build or fetch the effective stop words list.
+        Returns
+        -------
+        stop_words: list or None
+                A list of stop words.
+        """
+        if self.stop_words == "english":
+            return list(ENGLISH_STOP_WORDS)
+        elif isinstance(self.stop_words, str):
+            raise ValueError("not a built-in stop list: %s" % self.stop_words)
+        elif self.stop_words is None:
+            return None
+        else:  # assume it's a collection
+            return list(self.stop_words)
+
+    def get_char_ngrams(self, n, str_series, doc_id_sr):
+        """
+        Handles ngram generation for characters analyzers.
+
+        When analyzer is 'char_wb', we generate ngrams within word boundaries,
+        meaning we need to first tokenize and pad each token with a delimiter.
+        """
+        if self.analyzer == 'char':
+            token_count = str_series.str.len()
+            ngram_sr = str_series.str.character_ngrams(n=n)
+            ngram_count = None
+        else:
+            # here analyzer is 'char_wb'
+            tokens = str_series.str.tokenize(self.delimiter)
+
+            padding = Series(self.delimiter).repeat(len(tokens))
+            tokens = padding.str.cat(tokens.str.cat(padding))
+            tokens = tokens.reset_index(drop=True)
+
+            ngram_sr = tokens.str.character_ngrams(n=n)
+
+            token_count = str_series.str.token_count(self.delimiter)
+            doc_id_df = cudf.DataFrame({
+                'doc_id': doc_id_sr.repeat(token_count).reset_index(drop=True),
+                # formula to count ngrams given number of letters per token:
+                'ngram_count': tokens.str.len() - (n - 1)
+            })
+            ngram_count = doc_id_df.groupby('doc_id').sum()['ngram_count']
+
+        return ngram_sr, ngram_count, token_count
+
+    def get_ngrams(self, str_series, n, doc_id_sr):
+        """
+        This returns the ngrams for the string series
+        Parameters
+        ----------
+        str_series : (cudf.Series)
+            String series to tokenize
+        n : int
+            Gram level to get (1 for unigram, 2 for bigram etc)
+        doc_id_sr : cudf.Series
+            Int series containing documents ids
+        """
+        ngram_count = None
+        if self.analyzer == 'word':
+            token_count_sr = str_series.str.token_count(self.delimiter)
+            ngram_sr = str_series.str.ngrams_tokenize(n=n, separator=" ",
+                                                      delimiter=self.delimiter)
+        else:
+            if n != 1:
+                ngram_sr, ngram_count, token_count_sr = self.get_char_ngrams(
+                    n, str_series, doc_id_sr
+                )
+            else:
+                token_count_sr = str_series.str.len()
```

Comment:
I would probably move this `if` split down into the get_char_ngrams function. Sort of surprised me that this function didn't handle n=1, and since we have to do the work anyways, might as well push into the function and make it complete?

Reply:
Good idea.